File: baekjoon/bj9012.cpp
/*
문제:백준9012번
출처:https://www.acmicpc.net/problem/9012
*/
#include<iostream>
#include<stack>
#include<string>
using namespace std;

bool isVPS(string ps) {
	stack<char> s;
	for (int i = 0; i < ps.length(); i++) {
		if (ps[i] == '(') s.push(ps[i]);
		else {
			if (s.empty())return false;
			s.pop();
		}
	}
	if (!s.empty()) return false;
	return true;
}

int main() {
	int N;
	string ps;
	cin >> N;
	while (N--) {
		cin >> ps;
		if (isVPS(ps))
			cout << "YES" << '\n';
		else
			cout << "NO" << '\n';
	}
}

File: baekjoon/bj9012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class bj9012 {
    static Stack<Character> st;
    static int T;

    private static boolean isVPS(String ps) {
        st = new Stack<Character>();
        char[] tmp = ps.toCharArray();
        for(char c : tmp){
            if(c=='(')st.push(c);
            else if(c==')'){
                if(!st.isEmpty() && st.peek()=='(')st.pop();
                else return false;
            }
        }
        if(st.size()!=0)
            return false;

        else return true;
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer stk;
        T = Integer.parseInt(br.readLine());

        for(int i=0;i<T;i++){
            String ps = br.readLine();
            if(isVPS(ps)) System.out.println("YES");
            else System.out.println("NO");
        }

        br.close();
    }

}
